api/v1alpha1/applicationset_types.go

```diff
@@ -0,0 +1,73 @@
+package v1alpha1
+
+import (
+	"github.com/argoproj/argo-cd/pkg/apis/application/v1alpha1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// ApplicationSet is a set of Application resources
+// +kubebuilder:object:root=true
+// +genclient
+// +genclient:noStatus
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type ApplicationSet struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+	Spec              ApplicationSetSpec `json:"spec"`
+}
+
+// ApplicationSetSpec represents a class of application set state.
+type ApplicationSetSpec struct {
+	Generators ApplicationSetGenerators  `json:"generators"`
+	Template   ApplicationSetTemplate    `json:"template"`
+	Operation  *v1alpha1.SyncOperation   `json:"operation,omitempty"`
```

Comment:
Regarding the `operation` field:

1. Reusing the SyncOperation from the Applications APIs probably doesn't make sense here, since that object deals with a single git revision. But ApplicationSets may deal with many revisions, and even many repositories. If we do decide to have operation, we will need a new data structure.

2. In fact, we're not sure what a sync operation even means for an AppSet. Currently, I think it would only be used as a convenience field which carries forward setting the operation field for all the Applications in the ApplicationSet. But this could be done as a convenience in the API server instead.

3. We've had some regrets about introducing this operation field in Applications, since a better alternative might have been to introduce an `SyncOperation` CRD and controller to perform the logic. This CRD would act like a Job object and finish Successfully or Failed. 

So with the above, I think we should leave out operation, until we think about how we want to do operations (if at all).

Reply:
agree